panic: runtime error: invalid memory address or nil pointer dereference

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x77731d]

goroutine 1 [running]:
github.com/suzuki-shunsuke/deny-self-approve/pkg/github.(*Commit).Login(...)
	/home/runner/work/deny-self-approve/deny-self-approve/pkg/github/query.go:131
github.com/suzuki-shunsuke/deny-self-approve/pkg/controller.getCommitters(...)
	/home/runner/work/deny-self-approve/deny-self-approve/pkg/controller/run.go:92
github.com/suzuki-shunsuke/deny-self-approve/pkg/controller.(*Controller).Run(0xc00041b540, {0x8dfef0, 0xc0000a6500}, 0xc00018ab60, 0xc00041b580)
	/home/runner/work/deny-self-approve/deny-self-approve/pkg/controller/run.go:46 +0x49d
github.com/suzuki-shunsuke/deny-self-approve/pkg/cli.(*validateCommand).action(0xc000015170, 0xc0000a6880)
	/home/runner/work/deny-self-approve/deny-self-approve/pkg/cli/validate.go:59 +0x230
github.com/urfave/cli/v2.(*Command).Run(0xc0001bc000, 0xc0000a6880, {0xc000088c00, 0x2, 0x2})
	/home/runner/go/pkg/mod/github.com/urfave/cli/v2@v2.27.5/command.go:276 +0x7e2
github.com/urfave/cli/v2.(*Command).Run(0xc0001bcc60, 0xc0000a6600, {0xc0000140f0, 0x3, 0x3})
	/home/runner/go/pkg/mod/github.com/urfave/cli/v2@v2.27.5/command.go:269 +0xa65
github.com/urfave/cli/v2.(*App).RunContext(0xc0001ba000, {0x8dfef0, 0xc0000a6500}, {0xc0000140f0, 0x3, 0x3})
	/home/runner/go/pkg/mod/github.com/urfave/cli/v2@v2.27.5/app.go:333 +0x5a5
github.com/suzuki-shunsuke/deny-self-approve/pkg/cli.(*Runner).Run(0xc0000e3e68, {0x8dfef0, 0xc0000a6500})
	/home/runner/work/deny-self-approve/deny-self-approve/pkg/cli/runner.go:105 +0xc5d
main.core(0xc00018ab60)
	/home/runner/work/deny-self-approve/deny-self-approve/cmd/deny-self-approve/main.go:41 +0x1fe
main.main()
	/home/runner/work/deny-self-approve/deny-self-approve/cmd/deny-self-approve/main.go:22 +0x2f
```

--- a/README.md
+++ b/README.md
@@ -26,10 +26,10 @@ And it can also validate pull requests before doing something like deployment.
 
 [We provide a GitHub Actions to prevent self-approvals easily.](https://github.com/suzuki-shunsuke/deny-self-approve-action)
 
-## :warning: Commits not linking to GitHub Users
+## :warning: Commits not linked to GitHub Users
 
 deny-self-approve gets commits' users to deny self approvals.
-But if commits don't link to GitHub Users, deny-self-approve can't get committers.
+But if commits aren't linked to GitHub Users, deny-self-approve can't get committers.
 In that case, deny-self-approve fails by default.
 
 [You should configure Git properly](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user).

--- a/README.md
+++ b/README.md
@@ -32,8 +32,8 @@ deny-self-approve gets commits' users to deny self approvals.
 But if commits aren't linked to GitHub Users, deny-self-approve can't get committers.
 In that case, deny-self-approve fails by default.
 
-[You should configure Git properly](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user).
-[As a best practice, all commits should be verified](https://docs.github.com/en/authentication/managing-commit-signature-verification).
+- [You should configure Git properly](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user).
+- [As a best practice, all commits should be verified](https://docs.github.com/en/authentication/managing-commit-signature-verification).
 
 But if it's difficult, you can ignore these commits by `--ignore-not-linked-commit` option.
 

--- a/README.md
+++ b/README.md
@@ -26,10 +26,10 @@ And it can also validate pull requests before doing something like deployment.
 
 [We provide a GitHub Actions to prevent self-approvals easily.](https://github.com/suzuki-shunsuke/deny-self-approve-action)
 
-## :warning: Commits not linked to GitHub Users
+## :warning: Commit not linked to a GitHub User
 
 deny-self-approve gets commits' users to deny self approvals.
-But if commits aren't linked to GitHub Users, deny-self-approve can't get committers.
+But if a commit isn't linked to a GitHub User, deny-self-approve can't get a committer.
 In that case, deny-self-approve fails by default.
 
 - [You should configure Git properly](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user).

--- a/README.md
+++ b/README.md
@@ -25,3 +25,19 @@ And it can also validate pull requests before doing something like deployment.
 ## GitHub Actions
 
 [We provide a GitHub Actions to prevent self-approvals easily.](https://github.com/suzuki-shunsuke/deny-self-approve-action)
+
+## :warning: Commits not tied with GitHub Users
+
+deny-self-approve gets commits' users to deny self approvals.
+But if commits aren't tided with GitHub Users, deny-self-approve can't get committers.
+In that case, deny-self-approve fails by default.
+
+[As a best practice, all commits should be verified](https://docs.github.com/en/authentication/managing-commit-signature-verification).
+
+But if it's difficult, you can ignore these commits by `--ignore-unknown-commit` option.
+
+```sh
+deny-self-approve -ignore-unknown-commit validate # or dismiss
+```
+
+We don't recommend this because attackers can bypass the validation using unknown commits.

--- a/README.md
+++ b/README.md
@@ -30,15 +30,4 @@ And it can also validate pull requests before doing something like deployment.
 
 deny-self-approve gets commits' users to deny self approvals.
 But if a commit isn't linked to a GitHub User, deny-self-approve can't get a committer.
-In that case, deny-self-approve fails by default.
-
-- [You should configure Git properly](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user).
-- [As a best practice, all commits should be verified](https://docs.github.com/en/authentication/managing-commit-signature-verification).
-
-But if it's difficult, you can ignore these commits by `--ignore-not-linked-commit` option.
-
-```sh
-deny-self-approve --ignore-not-linked-commit validate # or dismiss
-```
-
-We don't recommend this because this isn't secure.
+In that case, deny-self-approve fails. [You need to configure Git properly](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user).

--- a/README.md
+++ b/README.md
@@ -26,18 +26,19 @@ And it can also validate pull requests before doing something like deployment.
 
 [We provide a GitHub Actions to prevent self-approvals easily.](https://github.com/suzuki-shunsuke/deny-self-approve-action)
 
-## :warning: Commits not tied with GitHub Users
+## :warning: Commits not linking to GitHub Users
 
 deny-self-approve gets commits' users to deny self approvals.
-But if commits aren't tided with GitHub Users, deny-self-approve can't get committers.
+But if commits don't link to GitHub Users, deny-self-approve can't get committers.
 In that case, deny-self-approve fails by default.
 
+[You should configure Git properly](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user).
 [As a best practice, all commits should be verified](https://docs.github.com/en/authentication/managing-commit-signature-verification).
 
-But if it's difficult, you can ignore these commits by `--ignore-unknown-commit` option.
+But if it's difficult, you can ignore these commits by `--ignore-not-linked-commit` option.
 
 ```sh
-deny-self-approve -ignore-unknown-commit validate # or dismiss
+deny-self-approve --ignore-not-linked-commit validate # or dismiss
 ```
 
-We don't recommend this because attackers can bypass the validation using unknown commits.
+We don't recommend this because this isn't secure.

--- a/pkg/cli/dismiss.go
+++ b/pkg/cli/dismiss.go
@@ -34,9 +34,10 @@ func (dc *dismissCommand) action(c *cli.Context) error {
 	gh.Init(c.Context, os.Getenv("GITHUB_TOKEN"))
 
 	input := &controller.Input{
-		PR:      c.Int("pr"),
-		Command: "dismiss",
-		Dismiss: true,
+		PR:                    c.Int("pr"),
+		Command:               "dismiss",
+		Dismiss:               true,
+		IgnoreNotLinkedCommit: c.Bool("ignore-not-linked-commit"),
 	}
 
 	if err := setRepo(c.String("repo"), input); err != nil {

--- a/pkg/cli/dismiss.go
+++ b/pkg/cli/dismiss.go
@@ -34,10 +34,9 @@ func (dc *dismissCommand) action(c *cli.Context) error {
 	gh.Init(c.Context, os.Getenv("GITHUB_TOKEN"))
 
 	input := &controller.Input{
-		PR:                    c.Int("pr"),
-		Command:               "dismiss",
-		Dismiss:               true,
-		IgnoreNotLinkedCommit: c.Bool("ignore-not-linked-commit"),
+		PR:      c.Int("pr"),
+		Command: "dismiss",
+		Dismiss: true,
 	}
 
 	if err := setRepo(c.String("repo"), input); err != nil {

--- a/pkg/cli/dismiss.go
+++ b/pkg/cli/dismiss.go
@@ -2,13 +2,8 @@ package cli
 
 import (
 	"io"
-	"os"
 
 	"github.com/sirupsen/logrus"
-	"github.com/spf13/afero"
-	"github.com/suzuki-shunsuke/deny-self-approve/pkg/controller"
-	"github.com/suzuki-shunsuke/deny-self-approve/pkg/github"
-	"github.com/suzuki-shunsuke/deny-self-approve/pkg/log"
 	"github.com/urfave/cli/v2"
 )
 
@@ -27,28 +22,5 @@ func (dc *dismissCommand) command() *cli.Command {
 }
 
 func (dc *dismissCommand) action(c *cli.Context) error {
-	logE := dc.logE
-	log.SetLevel(c.String("log-level"), logE)
-	log.SetColor(c.String("log-color"), logE)
-	gh := &github.Client{}
-	gh.Init(c.Context, os.Getenv("GITHUB_TOKEN"))
-
-	input := &controller.Input{
-		PR:      c.Int("pr"),
-		Command: "dismiss",
-		Dismiss: true,
-	}
-
-	if err := setRepo(c.String("repo"), input); err != nil {
-		return err
-	}
-
-	// TODO Get a pull request number from a commit hash
-	if err := getParamFromEnv(input); err != nil {
-		return err
-	}
-
-	ctrl := &controller.Controller{}
-	ctrl.Init(afero.NewOsFs(), gh, dc.stdout, dc.stderr)
-	return ctrl.Run(c.Context, logE, input) //nolint:wrapcheck
+	return commonAction(c, dc.logE, "dismiss", dc.stdout, dc.stderr, true)
 }

--- a/pkg/cli/runner.go
+++ b/pkg/cli/runner.go
@@ -77,6 +77,10 @@ func (r *Runner) Run(ctx context.Context) error {
 				Name:  "pr",
 				Usage: "The pull request number",
 			},
+			&cli.BoolFlag{
+				Name:  "ignore-unknown-commit",
+				Usage: "Ignore commits whose committer is unknown",
+			},
 		},
 		EnableBashCompletion: true,
 		Commands: []*cli.Command{

--- a/pkg/cli/runner.go
+++ b/pkg/cli/runner.go
@@ -9,7 +9,10 @@ import (
 	"time"
 
 	"github.com/sirupsen/logrus"
+	"github.com/spf13/afero"
 	"github.com/suzuki-shunsuke/deny-self-approve/pkg/controller"
+	"github.com/suzuki-shunsuke/deny-self-approve/pkg/github"
+	"github.com/suzuki-shunsuke/deny-self-approve/pkg/log"
 	"github.com/suzuki-shunsuke/go-ci-env/v3/cienv"
 	"github.com/suzuki-shunsuke/urfave-cli-help-all/helpall"
 	"github.com/urfave/cli/v2"
@@ -148,4 +151,30 @@ func getParamFromEnv(input *controller.Input) error {
 	}
 
 	return nil
+}
+
+func commonAction(c *cli.Context, logE *logrus.Entry, cmd string, stdout, stderr io.Writer, dismiss bool) error {
+	log.SetLevel(c.String("log-level"), logE)
+	log.SetColor(c.String("log-color"), logE)
+	gh := &github.Client{}
+	gh.Init(c.Context, os.Getenv("GITHUB_TOKEN"))
+
+	input := &controller.Input{
+		PR:      c.Int("pr"),
+		Command: cmd,
+		Dismiss: dismiss,
+	}
+
+	if err := setRepo(c.String("repo"), input); err != nil {
+		return err
+	}
+
+	// TODO Get a pull request number from a commit hash
+	if err := getParamFromEnv(input); err != nil {
+		return err
+	}
+
+	ctrl := &controller.Controller{}
+	ctrl.Init(afero.NewOsFs(), gh, stdout, stderr)
+	return ctrl.Run(c.Context, logE, input) //nolint:wrapcheck
 }

--- a/pkg/cli/runner.go
+++ b/pkg/cli/runner.go
@@ -78,8 +78,8 @@ func (r *Runner) Run(ctx context.Context) error {
 				Usage: "The pull request number",
 			},
 			&cli.BoolFlag{
-				Name:  "ignore-unknown-commit",
-				Usage: "Ignore commits whose committer is unknown",
+				Name:  "ignore-not-linked-commit",
+				Usage: "Ignore commits not linked to GitHub Users",
 			},
 		},
 		EnableBashCompletion: true,

--- a/pkg/cli/runner.go
+++ b/pkg/cli/runner.go
@@ -77,10 +77,6 @@ func (r *Runner) Run(ctx context.Context) error {
 				Name:  "pr",
 				Usage: "The pull request number",
 			},
-			&cli.BoolFlag{
-				Name:  "ignore-not-linked-commit",
-				Usage: "Ignore commits not linked to GitHub Users",
-			},
 		},
 		EnableBashCompletion: true,
 		Commands: []*cli.Command{

--- a/pkg/cli/validate.go
+++ b/pkg/cli/validate.go
@@ -40,9 +40,10 @@ func (vc *validateCommand) action(c *cli.Context) error {
 	gh.Init(c.Context, os.Getenv("GITHUB_TOKEN"))
 
 	input := &controller.Input{
-		PR:      c.Int("pr"),
-		Command: "validate",
-		Dismiss: c.Bool("dismiss"),
+		PR:                  c.Int("pr"),
+		Command:             "validate",
+		Dismiss:             c.Bool("dismiss"),
+		IgnoreUnknownCommit: c.Bool("ignore-unknown-commit"),
 	}
 
 	if err := setRepo(c.String("repo"), input); err != nil {

--- a/pkg/cli/validate.go
+++ b/pkg/cli/validate.go
@@ -40,10 +40,10 @@ func (vc *validateCommand) action(c *cli.Context) error {
 	gh.Init(c.Context, os.Getenv("GITHUB_TOKEN"))
 
 	input := &controller.Input{
-		PR:                  c.Int("pr"),
-		Command:             "validate",
-		Dismiss:             c.Bool("dismiss"),
-		IgnoreUnknownCommit: c.Bool("ignore-unknown-commit"),
+		PR:                    c.Int("pr"),
+		Command:               "validate",
+		Dismiss:               c.Bool("dismiss"),
+		IgnoreNotLinkedCommit: c.Bool("ignore-not-linked-commit"),
 	}
 
 	if err := setRepo(c.String("repo"), input); err != nil {

--- a/pkg/cli/validate.go
+++ b/pkg/cli/validate.go
@@ -2,13 +2,8 @@ package cli
 
 import (
 	"io"
-	"os"
 
 	"github.com/sirupsen/logrus"
-	"github.com/spf13/afero"
-	"github.com/suzuki-shunsuke/deny-self-approve/pkg/controller"
-	"github.com/suzuki-shunsuke/deny-self-approve/pkg/github"
-	"github.com/suzuki-shunsuke/deny-self-approve/pkg/log"
 	"github.com/urfave/cli/v2"
 )
 
@@ -33,28 +28,5 @@ func (vc *validateCommand) command() *cli.Command {
 }
 
 func (vc *validateCommand) action(c *cli.Context) error {
-	logE := vc.logE
-	log.SetLevel(c.String("log-level"), logE)
-	log.SetColor(c.String("log-color"), logE)
-	gh := &github.Client{}
-	gh.Init(c.Context, os.Getenv("GITHUB_TOKEN"))
-
-	input := &controller.Input{
-		PR:      c.Int("pr"),
-		Command: "validate",
-		Dismiss: c.Bool("dismiss"),
-	}
-
-	if err := setRepo(c.String("repo"), input); err != nil {
-		return err
-	}
-
-	// TODO Get a pull request number from a commit hash
-	if err := getParamFromEnv(input); err != nil {
-		return err
-	}
-
-	ctrl := &controller.Controller{}
-	ctrl.Init(afero.NewOsFs(), gh, vc.stdout, vc.stderr)
-	return ctrl.Run(c.Context, logE, input) //nolint:wrapcheck
+	return commonAction(c, vc.logE, "validate", vc.stdout, vc.stderr, c.Bool("dismiss"))
 }

--- a/pkg/cli/validate.go
+++ b/pkg/cli/validate.go
@@ -40,10 +40,9 @@ func (vc *validateCommand) action(c *cli.Context) error {
 	gh.Init(c.Context, os.Getenv("GITHUB_TOKEN"))
 
 	input := &controller.Input{
-		PR:                    c.Int("pr"),
-		Command:               "validate",
-		Dismiss:               c.Bool("dismiss"),
-		IgnoreNotLinkedCommit: c.Bool("ignore-not-linked-commit"),
+		PR:      c.Int("pr"),
+		Command: "validate",
+		Dismiss: c.Bool("dismiss"),
 	}
 
 	if err := setRepo(c.String("repo"), input); err != nil {

--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -28,9 +28,10 @@ func (c *Controller) Init(fs afero.Fs, gh GitHub, stdout, stderr io.Writer) {
 }
 
 type Input struct {
-	RepoOwner string
-	RepoName  string
-	PR        int
-	Dismiss   bool
-	Command   string
+	RepoOwner           string
+	RepoName            string
+	Command             string
+	PR                  int
+	Dismiss             bool
+	IgnoreUnknownCommit bool
 }

--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -28,10 +28,10 @@ func (c *Controller) Init(fs afero.Fs, gh GitHub, stdout, stderr io.Writer) {
 }
 
 type Input struct {
-	RepoOwner           string
-	RepoName            string
-	Command             string
-	PR                  int
-	Dismiss             bool
-	IgnoreUnknownCommit bool
+	RepoOwner             string
+	RepoName              string
+	Command               string
+	PR                    int
+	Dismiss               bool
+	IgnoreNotLinkedCommit bool
 }

--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -28,10 +28,9 @@ func (c *Controller) Init(fs afero.Fs, gh GitHub, stdout, stderr io.Writer) {
 }
 
 type Input struct {
-	RepoOwner             string
-	RepoName              string
-	Command               string
-	PR                    int
-	Dismiss               bool
-	IgnoreNotLinkedCommit bool
+	RepoOwner string
+	RepoName  string
+	Command   string
+	PR        int
+	Dismiss   bool
 }

--- a/pkg/controller/run.go
+++ b/pkg/controller/run.go
@@ -45,7 +45,7 @@ func (c *Controller) Run(ctx context.Context, logE *logrus.Entry, input *Input) 
 	}
 	committers, err := getCommitters(commits)
 	if err != nil {
-		if !input.IgnoreUnknownCommit {
+		if !input.IgnoreNotLinkedCommit {
 			return err
 		}
 		logerr.WithError(logE, err).Warn("some commits don't have committer and author")

--- a/pkg/controller/run.go
+++ b/pkg/controller/run.go
@@ -45,10 +45,7 @@ func (c *Controller) Run(ctx context.Context, logE *logrus.Entry, input *Input) 
 	}
 	committers, err := getCommitters(commits)
 	if err != nil {
-		if !input.IgnoreNotLinkedCommit {
-			return err
-		}
-		logerr.WithError(logE, err).Warn("commit isn't linked to a GitHub User")
+		return err
 	}
 	// Checks if people other than committers approve the PR
 	selfApprovals, ok := check(pr.HeadRefOID, pr.Reviews.Nodes, committers)

--- a/pkg/controller/run.go
+++ b/pkg/controller/run.go
@@ -16,7 +16,7 @@ import (
 // It gets pull request reviews and committers via GitHub GraphQL API, and checks if people other than committers approve the PR.
 // If Dismiss is true, it dismisses the approval of committers.
 // If the PR isn't approved by people other than committers, it returns an error.
-func (c *Controller) Run(ctx context.Context, logE *logrus.Entry, input *Input) error {
+func (c *Controller) Run(ctx context.Context, logE *logrus.Entry, input *Input) error { //nolint:cyclop
 	// Get a pull request reviews and committers via GraphQL API
 	pr, err := c.gh.GetPR(ctx, input.RepoOwner, input.RepoName, input.PR)
 	if err != nil {
@@ -43,7 +43,13 @@ func (c *Controller) Run(ctx context.Context, logE *logrus.Entry, input *Input) 
 	for i, commit := range pr.Commits.Nodes {
 		commits[i] = commit.Commit
 	}
-	committers := getCommitters(commits)
+	committers, err := getCommitters(commits)
+	if err != nil {
+		if !input.IgnoreUnknownCommit {
+			return err
+		}
+		logerr.WithError(logE, err).Warn("some commits don't have committer and author")
+	}
 	// Checks if people other than committers approve the PR
 	selfApprovals, ok := check(pr.HeadRefOID, pr.Reviews.Nodes, committers)
 	if input.Dismiss {
@@ -86,12 +92,21 @@ type Approval struct {
 	ID    string
 }
 
-func getCommitters(commits []*github.Commit) map[string]struct{} {
+func getCommitters(commits []*github.Commit) (map[string]struct{}, error) {
 	committers := make(map[string]struct{}, len(commits))
+	failed := false
 	for _, commit := range commits {
-		committers[commit.Login()] = struct{}{}
+		user, err := commit.Login()
+		if err != nil {
+			failed = true
+			continue
+		}
+		committers[user] = struct{}{}
 	}
-	return committers
+	if failed {
+		return committers, errors.New("both commiter and author are null")
+	}
+	return committers, nil
 }
 
 // check checks if committers approve the pull request themselves.

--- a/pkg/controller/run.go
+++ b/pkg/controller/run.go
@@ -48,7 +48,7 @@ func (c *Controller) Run(ctx context.Context, logE *logrus.Entry, input *Input) 
 		if !input.IgnoreNotLinkedCommit {
 			return err
 		}
-		logerr.WithError(logE, err).Warn("some commits don't have committer and author")
+		logerr.WithError(logE, err).Warn("commits aren't linked to GitHub Users")
 	}
 	// Checks if people other than committers approve the PR
 	selfApprovals, ok := check(pr.HeadRefOID, pr.Reviews.Nodes, committers)

--- a/pkg/controller/run.go
+++ b/pkg/controller/run.go
@@ -48,7 +48,7 @@ func (c *Controller) Run(ctx context.Context, logE *logrus.Entry, input *Input) 
 		if !input.IgnoreNotLinkedCommit {
 			return err
 		}
-		logerr.WithError(logE, err).Warn("commits aren't linked to GitHub Users")
+		logerr.WithError(logE, err).Warn("commit isn't linked to a GitHub User")
 	}
 	// Checks if people other than committers approve the PR
 	selfApprovals, ok := check(pr.HeadRefOID, pr.Reviews.Nodes, committers)
@@ -104,7 +104,7 @@ func getCommitters(commits []*github.Commit) (map[string]struct{}, error) {
 		committers[user] = struct{}{}
 	}
 	if failed {
-		return committers, errors.New("both commiter and author are null")
+		return committers, errors.New("commit isn't linked to a GitHub User")
 	}
 	return committers, nil
 }

--- a/pkg/controller/run.go
+++ b/pkg/controller/run.go
@@ -16,7 +16,7 @@ import (
 // It gets pull request reviews and committers via GitHub GraphQL API, and checks if people other than committers approve the PR.
 // If Dismiss is true, it dismisses the approval of committers.
 // If the PR isn't approved by people other than committers, it returns an error.
-func (c *Controller) Run(ctx context.Context, logE *logrus.Entry, input *Input) error { //nolint:cyclop
+func (c *Controller) Run(ctx context.Context, logE *logrus.Entry, input *Input) error {
 	// Get a pull request reviews and committers via GraphQL API
 	pr, err := c.gh.GetPR(ctx, input.RepoOwner, input.RepoName, input.PR)
 	if err != nil {

--- a/pkg/controller/run.go
+++ b/pkg/controller/run.go
@@ -91,17 +91,12 @@ type Approval struct {
 
 func getCommitters(commits []*github.Commit) (map[string]struct{}, error) {
 	committers := make(map[string]struct{}, len(commits))
-	failed := false
 	for _, commit := range commits {
-		user, err := commit.Login()
-		if err != nil {
-			failed = true
-			continue
+		login := commit.Login()
+		if login == "" {
+			return committers, errors.New("commit isn't linked to a GitHub User")
 		}
-		committers[user] = struct{}{}
-	}
-	if failed {
-		return committers, errors.New("commit isn't linked to a GitHub User")
+		committers[login] = struct{}{}
 	}
 	return committers, nil
 }

--- a/pkg/github/query.go
+++ b/pkg/github/query.go
@@ -127,10 +127,13 @@ type PullRequestCommit struct {
 }
 
 func (c *Commit) Login() (string, error) {
-	if c.Committer.User != nil {
+	if c == nil {
+		return "", errors.New("commit isn't linked to a GitHub User")
+	}
+	if c.Committer != nil && c.Committer.User != nil {
 		return c.Committer.User.Login, nil
 	}
-	if c.Author.User != nil {
+	if c.Author != nil && c.Author.User != nil {
 		return c.Author.User.Login, nil
 	}
 	return "", errors.New("commit isn't linked to a GitHub User")

--- a/pkg/github/query.go
+++ b/pkg/github/query.go
@@ -133,7 +133,7 @@ func (c *Commit) Login() (string, error) {
 	if c.Author.User != nil {
 		return c.Author.User.Login, nil
 	}
-	return "", errors.New("commit isn't linked to GitHub Users")
+	return "", errors.New("commit isn't linked to a GitHub User")
 }
 
 type Commit struct {

--- a/pkg/github/query.go
+++ b/pkg/github/query.go
@@ -133,7 +133,7 @@ func (c *Commit) Login() (string, error) {
 	if c.Author.User != nil {
 		return c.Author.User.Login, nil
 	}
-	return "", errors.New("both commiter and author are null")
+	return "", errors.New("commit isn't linked to GitHub Users")
 }
 
 type Commit struct {

--- a/pkg/github/query.go
+++ b/pkg/github/query.go
@@ -1,5 +1,7 @@
 package github
 
+import "errors"
+
 /*
 query($owner: String!, $repo: String!, $pr: Int!) {
   repository(owner: $owner, name: $repo) {
@@ -124,11 +126,14 @@ type PullRequestCommit struct {
 	Commit *Commit `json:"commit"`
 }
 
-func (c *Commit) Login() string {
+func (c *Commit) Login() (string, error) {
 	if c.Committer.User != nil {
-		return c.Committer.User.Login
+		return c.Committer.User.Login, nil
 	}
-	return c.Author.User.Login
+	if c.Author.User != nil {
+		return c.Author.User.Login, nil
+	}
+	return "", errors.New("both commiter and author are null")
 }
 
 type Commit struct {

--- a/pkg/github/query.go
+++ b/pkg/github/query.go
@@ -1,7 +1,5 @@
 package github
 
-import "errors"
-
 /*
 query($owner: String!, $repo: String!, $pr: Int!) {
   repository(owner: $owner, name: $repo) {
@@ -126,30 +124,41 @@ type PullRequestCommit struct {
 	Commit *Commit `json:"commit"`
 }
 
-func (c *Commit) Login() (string, error) {
-	if c == nil {
-		return "", errors.New("commit isn't linked to a GitHub User")
-	}
-	if c.Committer != nil && c.Committer.User != nil {
-		return c.Committer.User.Login, nil
-	}
-	if c.Author != nil && c.Author.User != nil {
-		return c.Author.User.Login, nil
-	}
-	return "", errors.New("commit isn't linked to a GitHub User")
-}
-
 type Commit struct {
 	Committer *Committer `json:"committer"`
 	Author    *Committer `json:"author"`
+}
+
+func (c *Commit) Login() string {
+	if c == nil {
+		return ""
+	}
+	if login := c.Committer.Login(); login != "" {
+		return login
+	}
+	return c.Author.Login()
 }
 
 type Committer struct {
 	User *User `json:"user"`
 }
 
+func (c *Committer) Login() string {
+	if c == nil {
+		return ""
+	}
+	return c.User.GetLogin()
+}
+
 type User struct {
 	Login string `json:"login"`
+}
+
+func (u *User) GetLogin() string {
+	if u == nil {
+		return ""
+	}
+	return u.Login
 }
 
 type Reviewer struct {

--- a/pkg/github/query_test.go
+++ b/pkg/github/query_test.go
@@ -11,7 +11,6 @@ func TestCommit_Login(t *testing.T) {
 	data := []struct {
 		name   string
 		commit *github.Commit
-		isErr  bool
 		exp    string
 	}{
 		{
@@ -44,23 +43,13 @@ func TestCommit_Login(t *testing.T) {
 				Author:    &github.Committer{},
 				Committer: &github.Committer{},
 			},
-			exp:   "",
-			isErr: true,
+			exp: "",
 		},
 	}
 	for _, d := range data {
 		t.Run(d.name, func(t *testing.T) {
 			t.Parallel()
-			login, err := d.commit.Login()
-			if err != nil {
-				if !d.isErr {
-					t.Fatal(err)
-				}
-				return
-			}
-			if d.isErr {
-				t.Fatal("error must be returned")
-			}
+			login := d.commit.Login()
 			if login != d.exp {
 				t.Fatalf("wanted: %s, got: %s", d.exp, login)
 			}

--- a/pkg/github/query_test.go
+++ b/pkg/github/query_test.go
@@ -1,0 +1,69 @@
+package github_test
+
+import (
+	"testing"
+
+	"github.com/suzuki-shunsuke/deny-self-approve/pkg/github"
+)
+
+func TestCommit_Login(t *testing.T) {
+	t.Parallel()
+	data := []struct {
+		name   string
+		commit *github.Commit
+		isErr  bool
+		exp    string
+	}{
+		{
+			name: "author",
+			commit: &github.Commit{
+				Committer: &github.Committer{},
+				Author: &github.Committer{
+					User: &github.User{
+						Login: "foo",
+					},
+				},
+			},
+			exp: "foo",
+		},
+		{
+			name: "committer",
+			commit: &github.Commit{
+				Author: &github.Committer{},
+				Committer: &github.Committer{
+					User: &github.User{
+						Login: "bar",
+					},
+				},
+			},
+			exp: "bar",
+		},
+		{
+			name: "not linked commit",
+			commit: &github.Commit{
+				Author:    &github.Committer{},
+				Committer: &github.Committer{},
+			},
+			exp:   "",
+			isErr: true,
+		},
+	}
+	for _, d := range data {
+		t.Run(d.name, func(t *testing.T) {
+			t.Parallel()
+			login, err := d.commit.Login()
+			if err != nil {
+				if !d.isErr {
+					t.Fatal(err)
+				}
+				return
+			}
+			if d.isErr {
+				t.Fatal("error must be returned")
+			}
+			if login != d.exp {
+				t.Fatalf("wanted: %s, got: %s", d.exp, login)
+			}
+		})
+	}
+}

